javascript/ql/src/semmle/javascript/frameworks/ShellJS.qll

```diff
@@ -162,6 +162,13 @@ module ShellJS {
     override predicate isShellInterpreted(DataFlow::Node arg) { arg = getACommandArgument() }
 
     override predicate isSync() {none ()}
+
+    override DataFlow::Node getOptionsArg() {
+      result = getLastArgument() and
+      not result = getArgument(0) and
+      not result.getALocalSource() instanceof DataFlow::FunctionNode and // looks like callback
```

Comment:
> I assume you have encountered a problem when we did not have the checks?   

Yep. 

All of these command executions methods have variations of the same API: `exec(command[, options][, callback])`. 

The API is implemented using runtime detection of the types of the arguments, so we have to do something similar. 

> At the very least, it would be nice with an explicitly test that exercises these instanceof cases.

:+1: 
(I found a bug or two while making those tests)